### PR TITLE
Move generation of replication task logic to task generator

### PR DESCRIPTION
**Pull request by wxing1292**
**What changed?**
* Move history replication task generation logic to task generator
* Add generate last history replication task API

**Why?**
Code re-structure

**How did you test it?**
Existing test

**Potential risks**
N/A

**Is hotfix candidate?**
No

### CODE REVIEW

File: service/history/workflow/mutable_state_impl.go [hunk 1]
```diff
@@ -3594,6 +3594,14 @@ func (e *MutableStateImpl) AddTimerTasks(
 	e.InsertTimerTasks = append(e.InsertTimerTasks, timerTasks...)
 }
 
+// AddReplicationTasks append visibility tasks
+func (e *MutableStateImpl) AddReplicationTasks(
+	replicationTasks ...tasks.Task,
+) {
+
+	e.InsertVisibilityTasks = append(e.InsertReplicationTasks, replicationTasks...)
```
Comment (yiminc) [inline service/history/workflow/mutable_state_impl.go:3602]: VisibilityTasks?

**Review by yiminc — APPROVED** [review]
how to you test this?

**Comment by wxing1292** [comment]
> how to you test this?

existing functionality will be tested by integration tests. 
the new generate last history replication task needs more integration test, until we have the time to restructure mutable state

File: service/history/workflow/task_generator.go [hunk 2]
```diff
@@ -509,6 +521,58 @@ func (r *TaskGeneratorImpl) GenerateUserTimerTasks(
...
+) error {
+	firstEvent := events[0]
+	lastEvent := events[len(events)-1]
+	if firstEvent.GetVersion() != lastEvent.GetVersion() {
+		return serviceerror.NewInternal("TaskGeneratorImpl encounter contradicting versions")
+	}
+	version := firstEvent.GetVersion()
+
+	r.mutableState.AddReplicationTasks(&tasks.HistoryReplicationTask{
+		// TaskID is set by shard
```
Comment (paulnpdev) [inline service/history/workflow/task_generator.go:537]: this comment is confusing given that TaskID doesn't appear here anywhere.

**Review by paulnpdev — COMMENTED** [review]
nits

**Pull request merged**